src/pipeline_migration/migrate.py

```diff
@@ -361,12 +327,108 @@ def fetch_migration_file(image: str, digest: str) -> str | None:
     return None
 
 
-def migrate(upgrades: list[dict[str, Any]]) -> None:
+def migrate(upgrades: list[dict[str, Any]], migration_resolver: type["Resolver"]) -> None:
     """The core method doing the migrations
 
     :param upgrades: upgrades data, that follows the schema of Renovate template field ``upgrades``.
     :type upgrades: list[dict[str, any]]
     """
-    manager = TaskBundleUpgradesManager(upgrades)
+    manager = TaskBundleUpgradesManager(upgrades, migration_resolver)
     manager.resolve_migrations()
     manager.apply_migrations()
+
+
+class Resolver(ABC):
+    """Base class for resolving migrations"""
+
+    @abstractmethod
+    def _resolve_migrations(
+        self, dep_name: str, upgrades_range: list[QuayTagInfo]
+    ) -> Generator[TaskBundleMigration, Any, None]:
+        raise NotImplementedError
+
+    def resolve(self, tb_upgrades: list[TaskBundleUpgrade]) -> None:
+
```

Comment:
Could you add a docstring to describe what this method should do? The type signature by itself isn't enough in this case because the method just modifies the `tb_upgrades` list in-place

Reply:
Updated. PTAL.